scapy/layers/dot11.py

```diff
@@ -240,6 +240,184 @@ def mysummary(self):
         else:
             return ""
 
+    registered_ies = {}
+
+    @classmethod
+    def register_variant(cls):
+        cls.registered_ies[cls.ID.default] = cls
+
+    @classmethod
+    def dispatch_hook(cls, _pkt=None, *args, **kargs):
+        if _pkt:
+            id_ = orb(_pkt[0])
+            if id_ == 221:
+                a = orb(_pkt[2])
+                b = orb(_pkt[3])
+                c = orb(_pkt[4])
+                if a == 0x00 and b == 0x50 and c == 0xf2:
+                    # MS OUI
+                    type_ = orb(_pkt[5])
+                    if type_ == 0x01:
+                        # MS WPA IE
+                        return Dot11EltMicrosoftWPA
+                    else:
+                        return Dot11EltVendorSpecific
+                else:
+                    return Dot11EltVendorSpecific
+            else:
+                return cls.registered_ies.get(id_, cls)
+        return cls
+
+    def haslayer(self, cls):
+        if cls == "Dot11Elt":
+            if isinstance(self, Dot11Elt):
```

Comment:
I have updated the code according to your comments.

Regarding the `haslayer()` code, the equivalent test should also be removed in other modules with the same logic (`bgp.py`, `eap.py` ...). Do you want me to submit a PR ?

Reply:
```
>>> generic_dot11_elt = Dot11Elt()
>>> rsn_ie = Dot11EltRSN()
>>> generic_dot11_elt.haslayer(Dot11EltRSN)
True
>>> generic_dot11_elt[Dot11EltRSN]
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-4-5c663853f282> in <module>()
----> 1 generic_dot11_elt[Dot11EltRSN]

scapy/scapy/packet.pyc in __getitem__(self, cls)
    975             elif not isinstance(lname, bytes):
    976                 lname = repr(lname)
--> 977             raise IndexError("Layer [%s] not found" % lname)
    978         return ret
    979 

IndexError: Layer [Dot11EltRSN] not found
```


```
>>> eap = EAP()
>>> eap_tls = EAP_TLS()
>>> eap.haslayer(EAP_TLS)
True
>>> eap[EAP_TLS]
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-4-48a591224de5> in <module>()
----> 1 eap[EAP_TLS]

scapy-haslayer/scapy/scapy/packet.pyc in __getitem__(self, cls)
   1003             elif not isinstance(lname, bytes):
   1004                 lname = repr(lname)
-> 1005             raise IndexError("Layer [%s] not found" % lname)
   1006         return ret
   1007 

IndexError: Layer [EAP_TLS] not found
```

Actually, this test is necessary in order to prevent `haslayer()` from returning `True` in some cases, such as the ones given in these examples.